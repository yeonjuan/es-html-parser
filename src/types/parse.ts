import { DocumentNode } from "./node";
import { TokenAdapter } from "./token-adapter";
import { AnyToken } from "./token";

export type ParseResult = {
  ast: DocumentNode;
  tokens: AnyToken[];
};

export type TemplateSyntaxOption = {
  name: string;
  start: string;
  end: string;
};

export type Options = {
  tokenAdapter?: TokenAdapter;
  templateSyntax?: TemplateSyntaxOption[];
};
