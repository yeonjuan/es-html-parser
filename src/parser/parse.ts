import { constructTree } from "../tree-constructor";
import { tokenize } from "../tokenizer";
import { DocumentNode, Token } from "../types";
import { clearParent } from "../utils";

export function parse(html: string): { ast: DocumentNode; tokens: Token[] } {
  const { tokens } = tokenize(html, undefined);
  const { ast } = constructTree(tokens, undefined);
  clearParent(ast);
  return {
    ast,
    tokens,
  };
}
