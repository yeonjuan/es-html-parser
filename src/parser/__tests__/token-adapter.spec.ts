import { parse } from "../parse";

describe("parse", () => {
  it("basic", () => {
    expect(parse("<div>content</div>")).toMatchSnapshot();
  });

  it("token adapter", () => {
    expect(
      parse("<div>content</div>", {
        tokenAdapter: {
          finalizeRange(token) {
            return [token.range[0] + 1, token.range[1] + 1];
          },
          finalizeLocation(token) {
            return {
              start: {
                line: token.loc.start.line + 1,
                column: token.loc.start.column,
              },
              end: {
                line: token.loc.end.line + 1,
                column: token.loc.end.column,
              },
            };
          },
        },
      })
    ).toMatchSnapshot();
  });
});
