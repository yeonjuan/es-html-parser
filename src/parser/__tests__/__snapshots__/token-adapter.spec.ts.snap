// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse basic 1`] = `
{
  "ast": {
    "children": [
      {
        "attributes": [],
        "children": [
          {
            "loc": {
              "end": {
                "column": 12,
                "line": 1,
              },
              "start": {
                "column": 5,
                "line": 1,
              },
            },
            "range": [
              5,
              12,
            ],
            "type": "Text",
            "value": "content",
          },
        ],
        "close": {
          "loc": {
            "end": {
              "column": 18,
              "line": 1,
            },
            "start": {
              "column": 12,
              "line": 1,
            },
          },
          "range": [
            12,
            18,
          ],
          "type": "CloseTag",
          "value": "</div>",
        },
        "loc": {
          "end": {
            "column": 18,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": "div",
        "openEnd": {
          "loc": {
            "end": {
              "column": 5,
              "line": 1,
            },
            "start": {
              "column": 4,
              "line": 1,
            },
          },
          "range": [
            4,
            5,
          ],
          "type": "OpenTagEnd",
          "value": ">",
        },
        "openStart": {
          "loc": {
            "end": {
              "column": 4,
              "line": 1,
            },
            "start": {
              "column": 0,
              "line": 1,
            },
          },
          "range": [
            0,
            4,
          ],
          "type": "OpenTagStart",
          "value": "<div",
        },
        "range": [
          0,
          18,
        ],
        "selfClosing": false,
        "type": "Tag",
      },
    ],
    "loc": {
      "end": {
        "column": 18,
        "line": 1,
      },
      "start": {
        "column": 0,
        "line": 1,
      },
    },
    "range": [
      0,
      18,
    ],
    "type": "Document",
  },
  "tokens": [
    {
      "loc": {
        "end": {
          "column": 4,
          "line": 1,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "range": [
        0,
        4,
      ],
      "type": "OpenTagStart",
      "value": "<div",
    },
    {
      "loc": {
        "end": {
          "column": 5,
          "line": 1,
        },
        "start": {
          "column": 4,
          "line": 1,
        },
      },
      "range": [
        4,
        5,
      ],
      "type": "OpenTagEnd",
      "value": ">",
    },
    {
      "loc": {
        "end": {
          "column": 12,
          "line": 1,
        },
        "start": {
          "column": 5,
          "line": 1,
        },
      },
      "range": [
        5,
        12,
      ],
      "templates": [],
      "type": "Text",
      "value": "content",
    },
    {
      "loc": {
        "end": {
          "column": 18,
          "line": 1,
        },
        "start": {
          "column": 12,
          "line": 1,
        },
      },
      "range": [
        12,
        18,
      ],
      "type": "CloseTag",
      "value": "</div>",
    },
  ],
}
`;

exports[`parse token adapter 1`] = `
{
  "ast": {
    "children": [
      {
        "attributes": [],
        "children": [
          {
            "loc": {
              "end": {
                "column": 12,
                "line": 2,
              },
              "start": {
                "column": 5,
                "line": 2,
              },
            },
            "range": [
              6,
              13,
            ],
            "type": "Text",
            "value": "content",
          },
        ],
        "close": {
          "loc": {
            "end": {
              "column": 18,
              "line": 2,
            },
            "start": {
              "column": 12,
              "line": 2,
            },
          },
          "range": [
            13,
            19,
          ],
          "type": "CloseTag",
          "value": "</div>",
        },
        "loc": {
          "end": {
            "column": 18,
            "line": 2,
          },
          "start": {
            "column": 0,
            "line": 2,
          },
        },
        "name": "div",
        "openEnd": {
          "loc": {
            "end": {
              "column": 5,
              "line": 2,
            },
            "start": {
              "column": 4,
              "line": 2,
            },
          },
          "range": [
            5,
            6,
          ],
          "type": "OpenTagEnd",
          "value": ">",
        },
        "openStart": {
          "loc": {
            "end": {
              "column": 4,
              "line": 2,
            },
            "start": {
              "column": 0,
              "line": 2,
            },
          },
          "range": [
            1,
            5,
          ],
          "type": "OpenTagStart",
          "value": "<div",
        },
        "range": [
          1,
          19,
        ],
        "selfClosing": false,
        "type": "Tag",
      },
    ],
    "loc": {
      "end": {
        "column": 18,
        "line": 2,
      },
      "start": {
        "column": 0,
        "line": 1,
      },
    },
    "range": [
      0,
      19,
    ],
    "type": "Document",
  },
  "tokens": [
    {
      "loc": {
        "end": {
          "column": 4,
          "line": 2,
        },
        "start": {
          "column": 0,
          "line": 2,
        },
      },
      "range": [
        1,
        5,
      ],
      "type": "OpenTagStart",
      "value": "<div",
    },
    {
      "loc": {
        "end": {
          "column": 5,
          "line": 2,
        },
        "start": {
          "column": 4,
          "line": 2,
        },
      },
      "range": [
        5,
        6,
      ],
      "type": "OpenTagEnd",
      "value": ">",
    },
    {
      "loc": {
        "end": {
          "column": 12,
          "line": 2,
        },
        "start": {
          "column": 5,
          "line": 2,
        },
      },
      "range": [
        6,
        13,
      ],
      "templates": [],
      "type": "Text",
      "value": "content",
    },
    {
      "loc": {
        "end": {
          "column": 18,
          "line": 2,
        },
        "start": {
          "column": 12,
          "line": 2,
        },
      },
      "range": [
        13,
        19,
      ],
      "type": "CloseTag",
      "value": "</div>",
    },
  ],
}
`;
